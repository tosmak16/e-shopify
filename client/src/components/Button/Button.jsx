import classNames from 'classnames';

import styles from './Button.scss';

export const BUTTON_TYPES = {
	STANDARD: 'standard',
	WARNING: 'warning',
	DANGER: 'danger',
	WHITE: 'white',
	PLAIN: 'plain'
};

class Button extends React.PureComponent {
	getButtonTypeName = () => {
		const { buttonType } = this.props;
		return Object.values(BUTTON_TYPES).indexOf(buttonType) > 0 ? `button--${buttonType}` : 'button--standard';
	}


	getClasses = () => {
		const {
			className, ripple, bordered, disabled,
		} = this.props;
		const buttonTypeName = this.getButtonTypeName();
		const classes = [buttonTypeName, className];
		return classNames({
			'button--disabled': disabled, ripple, bordered,
		}, classes);
	}

	renderButtonText = () => {
		const { children } = this.props;
		return (children);
	}

	render() {
		const {
			disabled,
			onClick,
			icon,
			customStyles,
			className,
		} = this.props;

		const classes = this.getClasses();

		return (
			<button
				type="button"
				styleName={classes}
				className={classNames(customStyles, className)}
				onClick={onClick}
				disabled={disabled}
			>
				{icon && icon}
				{this.renderButtonText()}
			</button>
		);
	}
}

Button.propTypes = {
	buttonType: PropTypes.oneOf(Object.values(BUTTON_TYPES)),
	children: PropTypes.string.isRequired,
	onClick: PropTypes.func.isRequired,
	className: PropTypes.string,
	customStyles: PropTypes.string,
	disabled: PropTypes.bool,
	icon: PropTypes.node,
	ripple: PropTypes.bool,
	bordered: PropTypes.bool,
};

Button.defaultProps = {
	buttonType: BUTTON_TYPES.STANDARD,
	disabled: false,
	className: '',
	customStyles: '',
	icon: null,
	ripple: false,
	bordered: false,
};

export default CSSModules(Button, styles, { allowMultiple: true });
