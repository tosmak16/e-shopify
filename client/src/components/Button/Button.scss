@import "~variables";

.button {
    width: 100%;
    font-family: $main-font;
    font-weight: bold;
    height: 50px;
    font-weight: bold;
    background: black;
    cursor: pointer;
    color: #fff;
    border: none;
    border-radius: 5px;
    box-shadow: 0 10px 20px 0 rgba(0, 0, 0, 0.12);
    font-size: 14px;
    letter-spacing: 0;
    text-align: center;
    display: flex;
    justify-content: center;
    align-items: center;
    transition: all .5s;

    &:focus,
    &:active {
        outline: none;
    }



}

.button--disabled {
    cursor: default;
    opacity: 0.5;
    pointer-events: none;
}

.button--danger,
.button--warning,
.button--standard,
.button--white,
.button--plain {
    composes: button;
}

.button {
    &--standard {
        color: $white;
        background: orange;

        &:hover {
            background-color: black;
            box-shadow: -3px -1px 5px #00000059;

        }


        &:active {
            box-shadow: 0px 0px 2px #00000059;
        }


    }

    &--warning {
        background: rgb(204, 201, 38);
    }

    &--danger {
        background: $fast-car-red;
        color: $white;
    }

    &--plain {
        color: black;
        background: $white;
        box-shadow: none;
    }

    &--white {
        background: white;
        color: black;

        &:hover {
            box-shadow: 0 5px 20px 0 rgba(0, 0, 0, 0.12);
            cursor: pointer;
        }

        &:active {
            background-color: black;
            color: $white;
        }
    }
}


.ripple {
    background-position: center;
    transition: background 1s;
}

.ripple:hover {
    background: black radial-gradient(circle, transparent 1%, rgba(0, 0, 0, 0.1) 1%) center/15000%;
}

.ripple:active {
    background-color: black;
    background-size: 100%;
    transition: background 0s;
    color: $white;
}

.bordered {
    border-radius: 4px;
}