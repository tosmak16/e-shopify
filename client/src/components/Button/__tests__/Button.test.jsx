import Button, { BUTTON_TYPES } from '../Button';

describe('Button', () => {
	it('contains the correct text', () => {
		const tree = shallow(<Button onClick={() => {}}>Hello world!</Button>);
		expect(tree.text()).toEqual('Hello world!');
	});
	it('emits correct action upon click', () => {
		let pass = false;
		const tree = shallow(<Button onClick={() => { pass = true; }}>Hello world!</Button>);
		tree.find('button').simulate('click');
		expect(pass).toEqual(true);
	});

	it('renders with icon', () => {
		const tree = shallow(<Button icon="icon text" onClick={() => { }}>Hello world!</Button>);
		expect(tree).toMatchSnapshot();
	});

	it('renders with other button type', () => {
		const tree = shallow(<Button buttonType={BUTTON_TYPES.PLAIN} icon="icon text" onClick={() => { }}>Hello world!</Button>);
		expect(tree).toMatchSnapshot();
	});
});
