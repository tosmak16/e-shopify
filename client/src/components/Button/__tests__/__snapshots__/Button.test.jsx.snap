// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders with icon 1`] = `
<button
  className="button--standard"
  disabled={false}
  onClick={[Function]}
  type="button"
>
  icon text
  Hello world!
</button>
`;

exports[`Button renders with other button type 1`] = `
<button
  className="button--plain"
  disabled={false}
  onClick={[Function]}
  type="button"
>
  icon text
  Hello world!
</button>
`;
